pkg/specs/pvcs.go

```diff
@@ -178,84 +184,123 @@ pvcLoop:
 			continue
 		}
 
+		// Detect the instance serial number.
+		// If it returns an error the PVC is ill-formed and we ignore it
+		serial, err := GetNodeSerial(pvc.ObjectMeta)
+		if err != nil {
+			continue
+		}
+		instances[serial] = append(instances[serial], pvc)
+
+		// Given that we are iterating over the PVCs
+		// we take the chance to build the list of resizing PVCs
 		if isResizing(pvc) {
 			result.Resizing = append(result.Resizing, pvc.Name)
 		}
+	}
 
-		// Find a Pod corresponding to this PVC
-		for idx := range podList {
-			if IsPodSpecUsingPVC(podList[idx].Spec, pvc.Name) {
-				// We found a Pod using this PVC so this
-				// PVC is not dangling
-				result.Healthy = append(result.Healthy, pvc.Name)
-				continue pvcLoop
-			}
+	// For every instance we have we validate the list of PVCs
+	// and detect if there is an attached Pod or Job
+instancesLoop:
+	for serial, pvcs := range instances {
+		instanceName := fmt.Sprintf("%s-%v", cluster.Name, serial)
+		expectedPVCs := getExpectedInstancePVCNames(cluster, instanceName)
+		pvcNames := getNamesFromPVCList(pvcs)
+
+		// If we have less PVCs that the expected number, all the instance PVCs are unusable
+		if len(expectedPVCs) > len(pvcNames) {
+			result.Unusable = append(result.Unusable, pvcNames...)
+			continue instancesLoop
 		}
 
-		for idx := range jobList {
-			if IsPodSpecUsingPVC(jobList[idx].Spec.Template.Spec, pvc.Name) {
-				// We have found a Job corresponding to this PVC, so we
-				// are initializing it or the initialization is just completed
-				result.Initializing = append(result.Initializing, pvc.Name)
-				continue pvcLoop
+		// If some PVC is missing, all the instance PVCs are unusable
+		for _, expectedPVC := range expectedPVCs {
+			if !slices.Contains(pvcNames, expectedPVC) {
+				result.Unusable = append(result.Unusable, pvcNames...)
+				continue instancesLoop
 			}
 		}
 
-		if pvc.Annotations[PVCStatusAnnotationName] != PVCStatusReady {
-			// This PVC has not a Job nor a Pod using it, but it is not marked as PVCStatusReady
-			// we need to ignore it here
-			continue
+		// If we have more PVC than expected, the extra PVCs are unusable
```

Comment:
If I understand correctly, what happens here is that we don't have "**more PVC than expected**" (as in count) but we found a PVC which we don't expect (as in subsets definition).

If I'm correct, perhaps we could fix this comment.

Reply:
Yes, you are right

Reply:
I agree